connector/src/main/java/com/datastax/oss/pulsar/source/converters/AbstractNativeConverter.java

```diff
@@ -0,0 +1,231 @@
+import com.datastax.oss.driver.api.core.metadata.schema.KeyspaceMetadata;
+import com.datastax.oss.driver.api.core.metadata.schema.TableMetadata;
+import com.datastax.oss.driver.api.core.type.DataType;
+import com.datastax.oss.driver.api.core.type.ListType;
+import com.datastax.oss.driver.api.core.type.MapType;
+import com.datastax.oss.driver.api.core.type.SetType;
+import com.datastax.oss.driver.api.core.type.UserDefinedType;
+import com.datastax.oss.protocol.internal.ProtocolConstants;
+import com.datastax.oss.pulsar.source.Converter;
+import lombok.extern.slf4j.Slf4j;
+import org.apache.avro.Schema;
+import org.apache.avro.SchemaBuilder;
+import org.apache.avro.generic.GenericRecord;
+import org.apache.pulsar.common.schema.SchemaType;
+
+import java.net.InetAddress;
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+
+/**
+ * An abstract class for native schema converters. Pulsar schemas (Avro and JSON) are represented in Avro with the only
+ * difference in the type field.
+ * @param <T> The desired type of the key representation. For example the Avro converter, the desired type is a List of
+ *           PK columns to because to query C* table. However, the actual keys are copied as is from the mutation topic.
+ *           In JSON only format, because the key will be embedded in the payload, the subclass may wish to convert to
+ *           pulsar's GenericRecord or serialized Jackson node
+ */
+@Slf4j
+public abstract class AbstractNativeConverter<T> implements Converter<byte[], GenericRecord, Row, T> {
+    public final org.apache.pulsar.client.api.Schema<byte[]> pulsarSchema;
+    public final Schema nativeSchema;
+    public final TableMetadata tableMetadata;
+    public final Map<String, Schema> subSchemas = new HashMap<>();
+
+    public AbstractNativeConverter(KeyspaceMetadata ksm, TableMetadata tm, List<ColumnMetadata> columns) {
+        this.tableMetadata = tm;
+        String keyspaceAndTable = ksm.getName() + "." + tm.getName();
+        List<Schema.Field> fields = new ArrayList<>();
+        for(ColumnMetadata cm : columns) {
+            boolean isPartitionKey = tm.getPartitionKey().contains(cm);
+            if (isSupportedCqlType(cm.getType())) {
+                Schema.Field field = fieldSchema(ksm, cm.getName().toString(), cm.getType(), !isPartitionKey);
+                if (field != null) {
+                    fields.add(field);
+                    switch(cm.getType().getProtocolCode()) {
+                        case ProtocolConstants.DataType.LIST:
+                        case ProtocolConstants.DataType.SET:
+                        case ProtocolConstants.DataType.MAP:
+                            Schema collectionSchema = dataTypeSchema(ksm, cm.getType());
+                            subSchemas.put(field.name(), collectionSchema);
+                            log.info("Add collection schema {}={}", field.name(), collectionSchema);
```

Comment:
break?

Reply:
it is optional in this context but better to add for readability.